From b7073b6f362d03df6a7b68f7760a57106e9ea67c Mon Sep 17 00:00:00 2001
From: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Date: Wed, 30 Oct 2013 15:54:01 +0000
Subject: [PATCH 035/351] ASoC: arizona: Add trigger callback for Ez2Control

All users to receive notification of the trigger being recieved in their
machine drivers.

Change-Id: I894cb25f6420e93a7819d55e813d2ec539de2c53
Signed-off-by: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Signed-off-by: Nikesh Oswal <Nikesh.Oswal@wolfsonmicro.com>
---
 include/linux/mfd/arizona/pdata.h |    3 +++
 sound/soc/codecs/arizona.c        |   11 +++++++++++
 sound/soc/codecs/arizona.h        |    2 ++
 sound/soc/codecs/wm5110.c         |    4 ++++
 4 files changed, 20 insertions(+)

diff --git a/include/linux/mfd/arizona/pdata.h b/include/linux/mfd/arizona/pdata.h
index 81e26be..5ecb07e 100644
--- a/include/linux/mfd/arizona/pdata.h
+++ b/include/linux/mfd/arizona/pdata.h
@@ -197,6 +197,9 @@ struct arizona_pdata {
 
 	/** General purpose switch control */
 	int gpsw;
+
+	/** Callback which is called when the trigger phrase is detected */
+	void (*ez2ctrl_trigger)(void);
 };
 
 #endif
diff --git a/sound/soc/codecs/arizona.c b/sound/soc/codecs/arizona.c
index 3fbeeda..cd31637 100644
--- a/sound/soc/codecs/arizona.c
+++ b/sound/soc/codecs/arizona.c
@@ -1775,6 +1775,17 @@ int arizona_set_hpdet_cb(struct snd_soc_codec *codec,
 }
 EXPORT_SYMBOL_GPL(arizona_set_hpdet_cb);
 
+int arizona_set_ez2ctrl_cb(struct snd_soc_codec *codec,
+			   void (*ez2ctrl_trigger)(void))
+{
+	struct arizona *arizona = dev_get_drvdata(codec->dev->parent);
+
+	arizona->pdata.ez2ctrl_trigger = ez2ctrl_trigger;
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(arizona_set_ez2ctrl_cb);
+
 MODULE_DESCRIPTION("ASoC Wolfson Arizona class device support");
 MODULE_AUTHOR("Mark Brown <broonie@opensource.wolfsonmicro.com>");
 MODULE_LICENSE("GPL");
diff --git a/sound/soc/codecs/arizona.h b/sound/soc/codecs/arizona.h
index c8ae40e..e35d0eb 100644
--- a/sound/soc/codecs/arizona.h
+++ b/sound/soc/codecs/arizona.h
@@ -257,5 +257,7 @@ int arizona_set_output_mode(struct snd_soc_codec *codec, int output,
 
 extern int arizona_set_hpdet_cb(struct snd_soc_codec *codec,
 				void (*hpdet_cb)(unsigned int));
+extern int arizona_set_ez2ctrl_cb(struct snd_soc_codec *codec,
+				  void (*ez2ctrl_trigger)(void));
 
 #endif
diff --git a/sound/soc/codecs/wm5110.c b/sound/soc/codecs/wm5110.c
index 7bf058c..1e331e1 100644
--- a/sound/soc/codecs/wm5110.c
+++ b/sound/soc/codecs/wm5110.c
@@ -1653,6 +1653,10 @@ static irqreturn_t adsp2_irq(int irq, void *data)
 
 	mutex_lock(&wm5110->compr_info.lock);
 
+	if (wm5110->core.arizona->pdata.ez2ctrl_trigger &&
+	    !wm5110->compr_info.total_copied)
+		wm5110->core.arizona->pdata.ez2ctrl_trigger();
+
 	ret = wm_adsp_stream_capture(wm5110->compr_info.adsp);
 	if (ret < 0) {
 		dev_err(wm5110->core.arizona->dev,
-- 
1.7.9.5

